### ISSUE TITLE
Dev: support for Blender 2.8x?

### ISSUE THREAD

**Issue by rafcolson** [GoranM/bdx#1041]
A question I was asking myself. I probably want to continue with Blender 2.8 for rendering 3d graphics and animations. But I'm afraid that it will be difficult to add support for BDX. A lot will be changed: layers are replaced by collections and I suppose that a lot of "game" related code is omitted.

So what do you guys think?

**Comment by GoranM** [comment]
Unless they're changing the core data model of how the blender python api works, I don't think it should be too difficult to refactor the addon to pull data from different sources.

**Comment by rafcolson** [comment]
I also think it should be possible, but if it would require a lot to be refactored, would it be worth it? Why not stick with Blender versions prior to 2.8?

**Comment by GoranM** [comment]
I don't really see how it would require "a lot to be refactored", because (as far as I understand) we would still export the same data, mostly from the same sources.

**Comment by rafcolson** [comment]
I was under the impression that with the coming of "Eevee" engine the "Blender Game" engine will be omitted in the official 2.8 release. But it seems that for now the test builds are still containing it, along with the API.

Time will tell.

**Comment by Gerkinfeltser** [comment]
Sounds like both the game engine & Blender internal render has been completely removed recently [in 2.8].

**Comment by SolarLune** [comment]
Yeah, I was thinking about this too - no idea how 2.8's API would be, or how "game-like" it'll be (with physics properties to tweak, for example). Maybe we'd basically have to start over for 2.8~

**Comment by rafcolson** [comment]
I had a look at what's been removed already and for example `bpy.types.Object.game` (`bpy.types.GameObjectSettings`) doesn't exist anymore. This also means that a lot of the menus which were essential to be able to work with BDX, are gone.

It's not for right now, but I think the best approach for compatibility with multiple versions (2.6-2.7-2.8) will be to create a separate Property tab, especially for BDX. Luckily Custom Properties (the properties on  `bpy.types.Object` are still available as was expected, so we can use these to create the necessary handles.

Note that with 2.8 we will also lose a lot of other useful functions, like copying object game properties and physics properties (from one object to other objects). So we might consider to reintroduce those functions in the Dynamic Spacebar Menu for example.

**Comment by rafcolson** [comment]
Hi guys,

It's been a long time. Hope you are doing fine.

Blender 2.80 is out now. I wonder if anyone is in for upgrading BDX for use with new releases of Blender. This project is as good as dead as no one is working on it anymore. But I might want to pursue an upgrade, only if it wouldn't be a waste of time for anyone to evaluate it. @GoranM @SolarLune 

Raf

**Comment by GoranM** [comment]
I received an email from SolarLune, a few months ago, expressing some interest in restarting BDX development. As I told him, I have no remaining interest in BDX, but if you guys want to fork it, and continue work on your own, feel free to do so.

That said: The big problem with LibGDX, and therefore BDX, is (in my somewhat humble opinion) layers upon layers of abstraction and dependencies, which are all fairly clunky, and which eventually create points of friction that become increasingly difficult to manage. With that in mind, I doubt that investing in BDX further is going to yield the kind of results you ultimately expect, and I think that you would probably be better served (especially in the long term) by something that is fundamentally simpler.

I think raylib (https://www.raylib.com/) is a fairly simple, and robust platform, which can be used to develop increasingly interesting games.

**Comment by rafcolson** [comment]
Thank you, Goran, especially for being so honest.

I agree that BDX is difficult to manage but it is a gem, nevertheless.

I'll still have to consider if continuing on my own (or with SolarLune) is worth the effort and time. I've had such fun with it in the past, learning in the process...

Thanks for all of your guidance over the years. I do consider you to be a good teacher.

Good luck on your future endeavors!

Raf

**Comment by SolarLune** [comment]
Hello!

I've been thinking about this for the past few days now, and I might be down to do something with this and Blender 2.8. 

I'm also wondering if it would be a good idea to try porting BDX in its current state, or possibly create a new engine from scratch... It might seem insane, but a lot of what we all did was manual work (setting transforms, exporting vertices, etc). I feel like if we took advantage of a more automatic approach afforded by Blender's existing exporters (if possible), the engine could be even smaller and simpler while also being more well-developed.

It might also be possible to recreate it using a custom engine entirely; what if, say, we were to create an engine using something else (C + Raylib, Haxe + Some existing 3D backend) as the backend instead of LibGDX? 

It'd be nice to be able to make the game and the engine in a single language to keep engine development simple as well, if possible...

**Comment by BluePrintRandom** [comment]
why not help upbge_eevee ?

https://blenderartists.org/t/upbge-eevee/1174109/5

https://github.com/UPBGE/blender/tree/eevee_merge8

can eevee be made to use openGLES?

**Comment by SolarLune** [comment]
Hello!

I'm not really interested in working with UPBGE - it has all of the issues of the original BGE, unfortunately (a huge codebase that one would have to learn before contributing anything concrete for engine devs, GPL licensing, and limited platform support for users). Improved graphical capabilities are nice, but portability and ease of use are more important to focus on, I believe.

Not a whole lot is to be gained from directly integrating the game engine inside of Blender's codebase in terms of core usability for the game engine.

**Comment by BluePrintRandom** [comment]
You can use bpy in game and do anything.

**Comment by SolarLune** [comment]
Yeah, I thought I read about that, but I'm not really sold on it making a huge difference in terms of actual results in the process of game development. It sounds like it opens up possibilities, but a "vanilla" game engine still has lots of possibilities, as well. I'm also not sure how having access to bpy would interact with the exported runtime; would the runtime effectively include all of Blender to facilitate bpy? I feel like it might be a good fit if you needed to make a complex modeling game, where you need access to the power of a 3D modeler to create the game tools that the player interacts with; in that case, a ton of work would be done for you. Otherwise, it's a ton of extra stuff that most devs probably won't use for their relatively small projects.

In any case, the core problems are still there, and those are things that are, in my opinion, more important than the benefits of being able to use Blender in-game.

**Comment by rafcolson** [comment]
@SolarLune

If we're considering to publish upgrades of BDX fot Blender 2.8x versions, then it would be proper to fork current master along with the wiki, so users wouldn't have to switch from one repo to another.

We then could decide if we want to continue development for enhancement, new features etc.

The main question I'm having now is who will be hosting BDX+. I'm willing to do that if that's okay for you.

**Comment by SolarLune** [comment]
That sounds fine.

**Comment by vittfiles** [comment]
Hello, I was working in a version from Blender 2.9 and this version have some news like g3db export, armature support and shadows.
this is the insall file:
[bdx.zip](https://github.com/GoranM/bdx/files/5767447/bdx.zip)
this is the code to compile:
https://drive.google.com/file/d/1bQVBpGmhfIf16FkbR6LQenLJNNx_LsTH/view?usp=sharing
install in preferences and create the proyect like the original version from BDX.
when the proyect files are create press run to download the jars.
then press export and run.
blender 2.8 had too many changes to physics editor and later I have to crate a wiki or sonthing to explain how to use it for bdx

.

**Comment by BluePrintRandom** [comment]
upbge has all that stuff restored - you may want to consider basing on it

**Comment by rafcolson** [comment]
@vittfiles Thanks for your effort in making those changes. I'm curious about it. Thanks!
